File: 백준/Gold/11437. LCA/LCA.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    private static List<List<Integer>> tree;
    private static int[] depth;
    private static int[] parent;
    private static boolean[] visited;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        tree = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            tree.add(new ArrayList<>());
        }

        for (int i = 0; i < n - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            tree.get(a).add(b);
            tree.get(b).add(a);
        }

        depth = new int[n + 1];
        parent = new int[n + 1];
        visited = new boolean[n + 1];

        bfs(1);

        int m = Integer.parseInt(br.readLine());
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < m; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());

            int lca = findLCA(a, b);
            sb.append(lca).append('\n');
        }

        System.out.print(sb);
    }

    private static void bfs(int root) {
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(root);
        visited[root] = true;
        depth[root] = 0;
        parent[root] = 0;

        while (!queue.isEmpty()) {
            int cur = queue.poll();

            for (int next : tree.get(cur)) {
                if (!visited[next]) {
                    visited[next] = true;
                    depth[next] = depth[cur] + 1;
                    parent[next] = cur;
                    queue.offer(next);
                }
            }
        }
    }

    private static int findLCA(int a, int b) {
        if (depth[a] < depth[b]) {
            int tmp = a;
            a = b;
            b = tmp;
        }

        while (depth[a] != depth[b]) {
            a = parent[a];
        }

        while (a != b) {
            a = parent[a];
            b = parent[b];
        }

        return a;
    }
}

File: 백준/Gold/11437. LCA/LCA.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.LinkedList
import java.util.Queue

private lateinit var tree: MutableList<MutableList<Int>>
private lateinit var depth: IntArray
private lateinit var parent: IntArray
private lateinit var visited: BooleanArray

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()

    tree = MutableList(n + 1) { mutableListOf() }

    repeat(n - 1) {
        val (a, b) = br.readLine().split(" ").map { it.toInt() }
        tree[a].add(b)
        tree[b].add(a)
    }

    depth = IntArray(n + 1)
    parent = IntArray(n + 1)
    visited = BooleanArray(n + 1)

    bfs(1)

    val m = br.readLine().toInt()
    val sb = StringBuilder()
    repeat(m) {
        val (a, b) = br.readLine().split(" ").map { it.toInt() }
        val lca = findLCA(a, b)
        sb.append(lca).append('\n')
    }

    print(sb)
}

private fun bfs(root: Int) {
    val queue: Queue<Int> = LinkedList()
    queue.offer(root)
    visited[root] = true
    depth[root] = 0
    parent[root] = 0

    while (queue.isNotEmpty()) {
        val cur = queue.poll()

        for (next in tree[cur]) {
            if (!visited[next]) {
                visited[next] = true
                depth[next] = depth[cur] + 1
                parent[next] = cur
                queue.offer(next)
            }
        }
    }
}

private fun findLCA(a: Int, b: Int): Int {
    var nodeA = a
    var nodeB = b

    if (depth[nodeA] < depth[nodeB]) {
        nodeA = nodeB.also { nodeB = nodeA }
    }

    while (depth[nodeA] != depth[nodeB]) {
        nodeA = parent[nodeA]
    }

    while (nodeA != nodeB) {
        nodeA = parent[nodeA]
        nodeB = parent[nodeB]
    }

    return nodeA
}
